tortoise/model/core.go

```diff
@@ -0,0 +1,200 @@
+		tortoise.WithLogger(logger.Named("trtl")),
+		tortoise.WithConfig(cfg),
+	)
+	return c
+}
+
+// core state machine.
+type core struct {
+	id     string
+	logger log.Log
+	rng    *rand.Rand
+
+	meshdb   *mesh.DB
+	atxdb    *activation.DB
+	beacons  *beaconStore
+	tortoise *tortoise.Tortoise
+
+	// generated on setup
+	units  uint32
+	signer signing.Signer
+
+	// set in the first layer of each epoch
+	refBallot     *types.BallotID
+	weight        uint64
+	eligibilities uint32
+
+	// set at the end of the epoch (MessageLayerEnd for the last layer of the epoch)
+	atx types.ATXID
+}
+
+// OnMessage receive blocks, atx, input vector, beacon, coinflip and store them.
+// Generate atx at the end of each epoch.
+// Generate block at the start of every layer.
+func (c *core) OnMessage(m Messenger, event Message) {
+	switch ev := event.(type) {
+	case MessageLayerStart:
+		// TODO(dshulyak) produce ballot according to eligibilities
+		if !ev.LayerID.After(types.GetEffectiveGenesis()) {
+			return
+		}
+		if c.refBallot == nil {
+			total, _, err := c.atxdb.GetEpochWeight(ev.LayerID.GetEpoch())
+			if err != nil {
+				panic(err)
+			}
+			c.eligibilities = max(uint32(c.weight*50/total), 1)
+		}
+		votes, err := c.tortoise.EncodeVotes(context.TODO())
+		if err != nil {
+			panic(err)
+		}
+		ballot := &types.Ballot{}
+		ballot.LayerIndex = ev.LayerID
+		ballot.Votes = *votes
+		ballot.AtxID = c.atx
+		for i := uint32(0); i < c.eligibilities; i++ {
+			ballot.EligibilityProofs = append(ballot.EligibilityProofs, types.VotingEligibilityProof{J: i})
+		}
+		if c.refBallot != nil {
+			ballot.RefBallot = *c.refBallot
+		} else {
+			_, activeset, err := c.atxdb.GetEpochWeight(ev.LayerID.GetEpoch())
+			if err != nil {
+				panic(err)
+			}
+			beacon, err := c.beacons.GetBeacon(ev.LayerID.GetEpoch())
+			if err != nil {
+				beacon = types.Beacon{}
+				c.rng.Read(beacon[:])
+				c.beacons.StoreBeacon(ev.LayerID.GetEpoch(), beacon)
+			}
+			ballot.EpochData = &types.EpochData{
+				ActiveSet: activeset,
+				Beacon:    beacon,
+			}
+		}
+		ballot.Signature = c.signer.Sign(ballot.Bytes())
+		ballot.Initialize()
+		if c.refBallot == nil {
+			id := ballot.ID()
+			c.refBallot = &id
+		}
+		m.Send(MessageBallot{Ballot: ballot})
+	case MessageLayerEnd:
+		if ev.LayerID.After(types.GetEffectiveGenesis()) {
+			verified := c.tortoise.HandleIncomingLayer(context.TODO(), ev.LayerID)
+			m.Notify(EventVerified{ID: c.id, Verified: verified, Layer: ev.LayerID})
+		}
+
+		if ev.LayerID.GetEpoch() == ev.LayerID.Add(1).GetEpoch() {
+			return
+		}
+
+		nipost := types.NIPostChallenge{
+			NodeID:     types.NodeID{Key: c.signer.PublicKey().String()},
+			StartTick:  1,
+			EndTick:    2,
+			PubLayerID: ev.LayerID,
+		}
+		atx := types.NewActivationTx(nipost, types.BytesToAddress(c.signer.PublicKey().Bytes()), nil, uint(c.units), nil)
+
+		c.refBallot = nil
+		c.atx = atx.ID()
+		c.weight = atx.GetWeight()
```

Comment:
i'd imagine these (refBallot/atx/weight) should be reset in MessageLayerStart event?